Add elementPlaceholder

jridgewell opened the pull request:
It's impossible to build large scale Backbone apps without composing
subviews into an overall parent view. The challenge with the normal
render-string->jQuery.html was preserving a subview's element inside the
re-rendered parent — `jQuery#html` wipes out all nodes (and events) and
resets everything in the parent.

```html
<!-- parent's template -->
<div id="subview" class="empty"></div>

<!-- subview's template -->
<span>Subview's content</span>

<!-- fully rendered content -->
<div>
  <div id="subview">
    <span>Subview's content</span>
  </div>
</div>
```

```js
// Rerender parent
parent.render();
```

```html
<div>
  <!-- subview just got wiped out
       it now references a detached node -->
  <div id="subview" class="empty"></div>
</div>
```

For the longest time, devs would get around this by either rendering,
then looping through all the subviews and calling `#setElement` with a
newly rendered DOM node (recreating event listeners), then re-render the
subview. Or the much faster way was to detach subviews first, re-render,
then swap the newly rendered DOM node with element already attached to
the subview.

```js
// Painfully slow
parent.render();
parent.subViews.forEach((subview) => {
  subview.setElement(parent.$(subview.selector));
  subview.render();
});

// Faster
parent.subViews.forEach((subview) => {
  subview.detach();
});
parent.render();
parent.subViews.forEach((subview) => {
  parent.$(subview.selector).replaceWith(subview.el);
});
```

Essentially, the parent view renders placeholders into its template and
the subviews grab (and possibly replace) the placeholder to do their
thing.

Incremental DOM works pretty similarly to the current Backbone patterns.
It will preserve any event handlers since they're bound to the subview's
root element. But, any children will be wiped away, and attributes will
be updated on the subview's element (that one's the subview's concern,
not the parent template's).

But, Incremental DOM has the potential to handle this case natively...

```js
elementOpen('div');
  elementPlaceholder('div', 'subview', ['id', 'subview']);
elementClose('div');
```

`elementPlaceholder` will act as just a placeholder. If it doesn't
exist in the parent, it will be created. If it does, it's left alone
giving the subview full autonomy over its element. No more
`#setElement`, rebinding event listeners, and re-rendering. And no more
detach-find-reattaches. :smile:

sparhami commented:
Somewhat along the lines of #57

We had been thinking about doing it using a specially named attribute before, but this approach works as well. One thing to would add would be a way to diff attributes on an arbitrary element, which might look like:

```javascript
var updateAttributes = IncrementalDOM.updateAttributes;

var el = elementPlaceholder('div', ...);
updateAttributes(el, 'name', value, ...);
```

One use for this would be to implement components on top of Incremental DOM without using webcomponents. There are a lot of ways that this could be done, but for the sake of example:

```javascript
function component(spec, props) {
  var el = elementPlaceholder(spec.tag, props.key);
  // possibly apply, say, everything in props.attributes to the element
  updateAttributes(el, ...);

  if (!el.initialized || spec.shouldUpdate.call(el, props, el.props)) {
    patch(el, spec.render.bind(el), props);
    el.initialized = true;
  }

  el.props = props;
}
```

An example component spec:

```javascript
var MyComponent = {
  tag: 'div',

  // called with 'this' as the element
  render: function(props) {
    // render the inside of the element
    text(props.text + ' ');
    props.content();
  },

  // called with 'this' as the element
  shouldUpdate: function(oldProps, newProps) {
    return oldProps.text !== newProps.text
  }
}
```

Using this example:

```javascript

elementOpen('div');
   component(MyComponent, {
     text: 'hello',
     content: function() {
       text('world');
     }
   });
elementClose('div');
```

ivan commented:
+1 on subview autonomy, I have an application with many rapidly-updated log windows (a thousand lines each) with where I think I want to use raw DOM calls to append text.

sparhami commented on test/functional/placeholder.js:
```diff
@@ -0,0 +1,95 @@
+    elementVoid = IncrementalDOM.elementVoid,
+    elementPlaceholder = IncrementalDOM.elementPlaceholder;
+
+describe('placeholder', () => {
```
Could you make these a bit more sentence like? If you combine each level of describes with the it, it should form a complete sentence. For example:

placeholders with no existing DOM element should render with the specified tag

jridgewell commented:
Updated the test descriptions and cleaned them up a bit.

> One thing to would add would be a way to diff attributes on an arbitrary element, which might look like

I can update #63 with that now that I know what you're looking for. :smile:

sparhami commented on src/virtual_elements.js:
```diff
@@ -75,6 +75,15 @@ if (!IS_PRODUCTION) {
+        throw new Error('Was expecting a key to be provided with the ' +
+            'placeholder.');
+      }
+  }
```
missing semicolon

sparhami commented on test/functional/placeholder.js:
```diff
@@ -0,0 +1,95 @@
+      expect(child.parentNode).to.equal(div);
+    });
+  });
+
```
extra blank line

sparhami commented on test/functional/placeholder.js:
```diff
@@ -0,0 +1,95 @@
+    it('should leave any decedents unaltered', () => {
+      var child = container.querySelector('#child');
+
+      expect(child.parentNode).to.equal(div);
```
This is too roundabout. For example, you don't know that the child is the same one initially rendered by reading the test case. Tests should be as explicit as possible.

sparhami commented on test/functional/placeholder.js:
```diff
@@ -0,0 +1,95 @@
+      });
+
+      patch(container, () => {
+        el = elementPlaceholder('span', 'placeholder', ['class', 'someClass'], 'data-foo', 'bar');
```
This is confusing binds us to a much stronger contract than we might want to have in the future. You should change the test to use `elementPlaceholder` and `div` in the initial render.

The functional tests should act as part of the spec of Incremental DOM and should be as restrictive as possible.

jridgewell replied:
> This is confusing binds us to a much stronger contract than we might want to have in the future. You should change the test to use elementPlaceholder and div in the initial render.

I won't be able to create the child element with `elementPlaceholder`. I can make it:

```js
      patch(container, function() {
        div = elementPlaceholder('div', 'placeholder');
        patch(div, function() {
          child = elementVoid('div', null, ['id', 'child']);
        });
        elementClose('div');
      });
```

sparhami replied:
You could also just do an innerHTML after the first render

sparhami commented on src/virtual_elements.js:
```diff
@@ -75,6 +75,18 @@ if (!IS_PRODUCTION) {
 
 
+  /**
+  * Makes sure that the caller provided a key for the placeholder
```
missing period

jridgewell commented:
Updated.

sparhami commented on test/functional/placeholder.js:
```diff
@@ -0,0 +1,94 @@
+      });
+
+      patch(container, () => {
+        el = elementPlaceholder('span', 'placeholder', ['class', 'someClass'], 'data-foo', 'bar');
```
The tags should agree, so this should be 'div'. I am very likely to add an assert to make sure something with the same key always has the same tag.

jridgewell replied:
I'm afraid that breaks my use case. I render all my placeholder elements as `span`s (so they collapse on the DOM) and replace the placeholder with the subview's element. But I can force the subview to be a `span` (same argument for `div`).

sparhami replied:
Would it be possible create a wrapping function that creates it with the correct tag and `display: inline` instead? Or maybe just manually pass it.

Having something with the same key but a different tag might be a user error. For example, if you had an array and did something like:

```
items.forEach((item) => {
  if (item.foo) {
    elementVoid('div', item.id);
  } else {
    elementVoid('span', item.id);
  }
});
```

You would expect that if the condition changes that the tag should change as well.

jridgewell replied:
> Would it be possible create a wrapping function that creates it with the correct tag and display: inline instead? Or maybe just manually pass it.

Subview's aren't known to the parent until one is inserted and, even then, another subview can replace it.

```html
<!-- Tab 1's template, including root -->
<ul>...</ul>
<!-- tab 1 -->

<!-- Tab 2's template, including root -->
<section><p>...</p></section>
<!-- tab 2 -->


<div id="parent">
  <nav><a href="#tab1">Tab 1</a> | <a href="#tab2">Tab 2</a></nav>
  <span key="placeholder"></span>
</div>
```

Wrapping divs would be the only generic solution here, but personally despise them.

> You would expect that if the condition changes that the tag should change as well.

For an `elementVoid`, yes. But I'm trying to define placeholders as generic identifiers that _just_ hold a position in the DOM tree, since it's impossible to track the position without inserting some element in that position. What's currently occupying that position shouldn't matter, as long as the position is maintained.

sparhami commented on test/functional/placeholder.js:
```diff
@@ -0,0 +1,94 @@
+    });
+
+    it('should not set dynamic attributes', () => {
+      expect(div.getAttribute('data-foo')).not.to.equal('bar');
```
There isn't a need to test things that are not part of the function signature.

jridgewell replied:
It's actually to be preemptive, since I imagine someone is going to request that placeholder's have their attributes updated. I would feel comfortable with just testing `statics` aren't updated (by creating a new placeholder element), but they don't update on normal elements anyways.

sparhami replied:
On second thought, I don't see a good reason not to allow for updating of attributes on placeholders. It would be strictly optional anyway as we don't remove attributes that have not been touched by the library.

jridgewell replied:
In theory, because it's not the parent's concern. The placeholder (or whatever is attached to the placeholder) should have full autonomy in changing the element. I believe React also follows this pattern:

```js
var Query = React.createClass({
  render: function() {
    return (
      // The "placeholder" element. Query has full autonomy.
      <td className={ "Query " + this.props.elapsedClassName}>
        {this.props.formatElapsed}
      </td>
    );
  }
})
// Usage. The attributes here aren't actually set on
// the element, but passed to Query so it may control.
<Query key={index}
  formatElapsed={query.formatElapsed}
  elapsedClassName={query.elapsedClassName} />
```

And drawing from my Backbone usecase above, the placeholder is wholly replaced by the subview's element. So even if `elementPlaceholder` were to add some dynamic attributes to the element, the element is just going to be swapped with some different element without those attributes.

sparhami replied:
Actually, I believe it is the parent's concern. The outside should be controlled by the one instantiating the component with the inside controlled by the component itself.

Take for example the '<select>' element. When you declare the element, you can give it some configuration options via attributes, for example `disabled` or `multiple`. You may want to dynamically change the `disabled` attribute based on some other form control. The select attribute has some internal components. For example, it has an arrow on the right, but that is internal to the component.

There are quite a few problems with the way React does things which I would like to avoid altogether. For example, you will see people passing classes to components, which will conditionally add it to a class that they themselves define. It ends up that many components do this, but you never know which ones do it and which ones don't without going into the component itself and checking. It also ends up being extremely verbose since you sometimes need to pass a large number of properties which in turn are applied to the root element.

For Backbone, I think it is fine swapping the whole placeholder. Others may not be using Backbone, so there is no reason to artificially limit the library for that reason.

jridgewell replied:
>  The outside should be controlled by the one instantiating the component with the inside controlled by the component itself.

The pattern of the outside controlling the container while the render function controls everything inside it doesn't quite apply anymore with the [recommendation](https://github.com/google/incremental-dom/pull/66#issuecomment-121853982) to pull out an `updateAttributes` function from general DOM alignment. 

You can't do a normal `elementOpen`/`alignWithDOM` on the `container` because then your container reference could point to a removed element. But `updateAttributes` would only mutate the attributes on the element, never removing it. Now we can treat the container as just a root element, and the view/component/instance attached to that root element is able to exist in isolation from a parent. That principle is one of the things that makes Backbone so damn nice: view's don't require a parent to exist and are fully autonomous and useful without one.

> For example, you will see people passing classes to components, which will conditionally add it to a class that they themselves define. It ends up that many components do this, but you never know which ones do it and which ones don't without going into the component itself and checking. 

You'd still be able to mutate a DOM node outside of the normal flow with `updateAttributes`, even inside the same `render`.

But I think breaking encapsulation by supporting this as a first class API is a much more harmful. Imagine some subview needs a `special` class for the correct CSS styles. But now, a parent comes along and tells the subview to remove it's class.

sparhami replied:
> The pattern of the outside controlling the container while the render function controls everything inside it doesn't quite apply anymore with the recommendation to pull out an updateAttributes function from general DOM alignment.

The point was to give control back to the library using Incremental DOM control, if it wants it. For example, you might not want to do any attribute value caching and take the performance hit in exchange for a reduced memory footprint.

After thinking about it, alignment and attribute diffing are really separate things so they probably shouldn't even be tied together.

> Now we can treat the container as just a root element, and the view/component/instance attached to that root element is able to exist in isolation from a parent. That principle is one of the things that makes Backbone so damn nice: view's don't require a parent to exist and are fully autonomous and useful without one.

In my opinion, the insertion point should not act as the root of a view. That would mean someone external to your view could decide that your root element should be an `h3` instead of a `div`.  And while I think doing that is not the right pattern, Incremental DOM will not enforce it.

> You'd still be able to mutate a DOM node outside of the normal flow with updateAttributes, even inside the same render.

I don't mind Incremental DOM giving people footguns, as it isn't intended to be used directly like that. The DOM API itself gives you plenty and we're just providing a thin wrapper on top of that.

> But I think breaking encapsulation by supporting this as a first class API is a much more harmful. Imagine some subview needs a special class for the correct CSS styles. But now, a parent comes along and tells the subview to remove it's class.

This comes from the fact that you are trying to use this insertion point that the parent should have control over in the subview. If you really want to retain control, the call to `elementPlaceholder` should be moved outside of the control of the parent. For example, instead of:

```javascript
elementOpen('div');
  elementPlaceholder('div', 'subview', ...);
elementClose('div);
```

you could do something like:

```javascript
elementOpen('div');
  subview({
    viewName: 'subview',
    ...
  });
elementClose('div');

// somewhere else
function subview(params) {
  elementPlaceholder('div', params.viewName, ...);
}
```

jridgewell replied:
Updated.

jridgewell commented:
I pinged you in https://github.com/babel-plugins/babel-plugin-incremental-dom/issues/9, but to continue the Incremental DOM issues here:

@thejameskyle and I are currently running into issues with the lifecycle of placeholder elements, specifically `create` and `destroy` events. We can use a `component` helper function to take over management of a placeholder with some `Component` class, but we have no way of knowing if the component needs to be created, has already been created, or is being removed from the view.

Thoughts on how we might tackle this?

jridgewell commented:
Closed by #120.
